Fix ODSP Fetch 412 in Summarizer E2E test:  Simply close the summarizer just before uploading.

NicholasCouri opened the pull request:
## Description

Closing after the upload would not work on ODSP as the summarizer handle would be returned, and the server would have this handle id stored as the last summarizer. The next summarizer, getting from cache, would fail.
We could have force to get from network, which would work, or we can simply forcibly fail to upload by closing the summarizer instead. Fix [#AB4075](https://dev.azure.com/fluidframework/internal/_workitems/edit/4075)

NicholasCouri committed:
Simply close the summarizer just before uploading.
Closing after the upload would not work for ODSP as the summarizer handle would be returned, it, the server will have this version stored and the next summarizer, getting from cache would fail.

NicholasCouri committed:
Merge branch 'main' into SummarizerFailures

vladsud commented:
@NicholasCouri, the test seems to be correct (before changes) as it hits specific condition that could happen in real life (and it validates how system works). It feels like we should not disable it, but search for other options if it makes things noisy in our service tests or end-to-end tests when using real service.
Maybe have a list of "expected" errors that we downgrade to generic events and thus do not bring attention to them?
Worse solution - mark the test to be run only in UTs, i.e. not involving real service (but I do not like it).

agarwal-navin commented on packages/test/test-end-to-end-tests/src/test/SummarizeFetchValidation.spec.ts:
```diff
@@ -369,10 +369,9 @@ describeFullCompat("Summarizer fetches expected number of times", (getTestObject
-			const response = await uploadSummaryUploaderFunc(summary, context);
 			// Close summarizer so that it does not submit SummaryOp
 			summarizer.close();
-			return response;
```
This seems hacky. If the purpose of the test is to just fail summaries, you could override the `summarize` function instead and just throw an error in it. You could also throw an error in this function itself without doing any of the other stuff. That would make it much clear what is happening.

NicholasCouri replied:
The goal was to simply close the summarizer before the upload happens. I could also close and let the upload happen. That would not affect anything.  Or I could special case for ODSP and force fetch from latest on the next summarize.  That would also do it.

NicholasCouri committed:
There are several additional errors that came from describeFullCompat as the expectaions for fetch are different per version.
I'm going back to NoCompat until we figure out the best way to deal with these differences.

NicholasCouri commented:
> @NicholasCouri, the test seems to be correct (before changes) as it hits specific condition that could happen in real life (and it validates how system works). It feels like we should not disable it, but search for other options if it makes things noisy in our service tests or end-to-end tests when using real service. Maybe have a list of "expected" errors that we downgrade to generic events and thus do not bring attention to them? Worse solution - mark the test to be run only in UTs, i.e. not involving real service (but I do not like it).

Yes, I reverted  to the original changes - the problem is that we have different behaviors on N, N-1 combinations. It is doable to have a set of expected values. If you think it is worthy it, I can be as granular as possible but it will require  additional effort.

NicholasCouri committed:
prettier:fix

agarwal-navin commented on packages/test/test-end-to-end-tests/src/test/SummarizeFetchValidation.spec.ts:
```diff
@@ -367,11 +367,17 @@ describeFullCompat("Summarizer fetches expected number of times", (getTestObject
 			summarizer.close();
+			// ODSP has single commit summary enabled by default and
+			// will update the summary version even without the summary op.
+			if (provider.driver.type === "odsp") {
```
Is this check needed? You could do this for all drivers right.
